src/bp/ui-studio/src/web/components/Layout/training-status-observer.ts

```diff
@@ -0,0 +1,87 @@
+import axios from 'axios'
+import { NLU } from 'botpress/sdk'
+import EventBus from '~/util/EventBus'
+
+type TrainingStatusListener = {
+  name: string
+  cb: (session: NLU.TrainingSession, fromWebSocket: boolean) => void
+  error: (err: Error) => void
+}
+
+class TrainingStatusObserver {
+  private listeners: TrainingStatusListener[] = []
+  private pollingHandle: number | undefined
+  private session: NLU.TrainingSession
+  private language: string
+
+  constructor() {}
+
+  setLanguage(language: string) {
+    this.language = language
+  }
+
+  dispose() {
+    this.stopPolling()
+  }
+
+  addListener(listener: TrainingStatusListener) {
+    if (!this.listeners.length) {
+      this.startPolling()
+    }
+    this.listeners.push(listener)
+  }
+
+  removeListener(listener: TrainingStatusListener) {
+    const idx = this.listeners.findIndex(l => l.name === listener.name)
+    if (idx > -1) {
+      this.listeners.splice(idx, 1)
+    }
+
+    if (!this.listeners.length) {
+      this.stopPolling()
+    }
+  }
+
+  private startPolling() {
+    EventBus.default.on('statusbar.event', this.onStatusBarEvent)
+    this.pollingHandle = window.setInterval(
+      () => this.session?.status !== 'training' && this.fetchTrainingStatus(),
+      1500
+    ) // for training-needed
+  }
+
+  private stopPolling() {
+    clearInterval(this.pollingHandle)
+    EventBus.default.off('statusbar.event', this.onStatusBarEvent)
+  }
+
+  private onStatusBarEvent = async event => {
+    const isNLUEvent = event.botId === window.BOT_ID && event.trainSession?.language === this.language
+    if (isNLUEvent) {
+      this.notifyListeners(event.trainSession as NLU.TrainingSession, false)
+    }
+  }
+
+  private notifyListeners(ts: NLU.TrainingSession, fromWebSocket: boolean) {
+    this.session = ts
+    for (const listener of this.listeners) {
```

Comment:
Just a preference, could be applied in notifyError too

```suggestion
    this.listeners.forEach(listener => {
```